tests/apitest/src/CoreServices/FSEventStream.cs

```diff
@@ -0,0 +1,177 @@
+		[TestFixtureTearDown]
+		public void Destroy ()
+		{
+			Directory.Delete ("FSEventStreamTests", true);
+		}
+
+		[SetUp]
+		public void CreateFSEventStream ()
+		{
+			watchedFileChanged = false;
+			exludedFileChanged = false;
+
+			fsEventStream = new FSEventStream (st_dev, pathsToWatchRelativeToDevice, FSEvent.SinceNowId, TimeSpan.FromSeconds (1), FSEventStreamCreateFlags.WatchRoot | FSEventStreamCreateFlags.FileEvents | FSEventStreamCreateFlags.NoDefer);
+			Assert.IsNotNull (fsEventStream, "Null");
+		}
+
+		[TearDown]
+		public void EndFSEventStream ()
+		{
+			fsEventStream.Stop ();
+			fsEventStream.Invalidate ();
+			fsEventStream.Dispose ();
+		}
+
+		[Test]
+		public void GetDeviceBeingWatchedTest ()
+			=> Assert.AreEqual (st_dev, fsEventStream.GetDevice (), "Device ID");
+
+		[Test]
+		public void FSEventFileChangedTest ()
+		{
+			string path = Path.Combine (dirPath, "TempFileToWatch.txt");
+			var taskCompletionSource = new TaskCompletionSource<FSEventStreamEventsArgs> ();
+			FSEventStreamEventsArgs args = null;
+
+			TestRuntime.RunAsync (TimeSpan.FromSeconds (30), async () => {
+				fsEventStream.Events += (sender, eventArgs) => {
+					taskCompletionSource.SetResult (eventArgs);
+					watchedFileChanged = true;
+				};
+				fsEventStream.ScheduleWithRunLoop (CFRunLoop.Current);
+				fsEventStream.Start ();
+				File.AppendAllText (path, "Hello World!");
+				Assert.IsTrue (File.Exists (path));
+				args = await taskCompletionSource.Task.ConfigureAwait (false);
+
+			}, () => watchedFileChanged);
+
+			Assert.IsNotNull (args, "Null args");
+		}
+
+		[Test]
+		public void UnscheduleFromRunLoopTest ()
+		{
+			string path = Path.Combine (dirPath, "TempFileToWatch.txt");
+			var taskCompletionSource = new TaskCompletionSource<FSEventStreamEventsArgs> ();
+			FSEventStreamEventsArgs args = null;
+
+			TestRuntime.RunAsync (TimeSpan.FromSeconds (30), async () => {
+				fsEventStream.Events += (sender, eventArgs) => {
+					taskCompletionSource.SetResult (eventArgs);
+					watchedFileChanged = true;
+				};
+				fsEventStream.UnscheduleFromRunLoop (CFRunLoop.Current); // unscheduling from the RunLoop shouldn't trigger events 
+				fsEventStream.Start ();
+				File.AppendAllText (path, "Hello World!");
+				Assert.IsTrue (File.Exists (path));
+				args = await taskCompletionSource.Task.ConfigureAwait (false);
+
+			}, () => watchedFileChanged);
+
+			Assert.IsNull(args, "Null args");
+		}
+
+		[Test]
+		public void SetExclusionPathsTest ()
+		{
+			var watchDirPath = Path.Combine (dirPath, "WatchDir");
+			var watchDir = Directory.CreateDirectory (watchDirPath);
+			var unWatchDirPath = Path.Combine (dirPath, "UnwatchDir");
+			var unWatchDir = Directory.CreateDirectory (unWatchDirPath);
+
+			// Excluding the unWatchDirPath from the watcher so any event inside it doesn't get triggered
+			Assert.IsTrue (fsEventStream.SetExclusionPaths (new [] { Path.Combine ("private", unWatchDirPath.Substring (1)) }), "SetExclusionPaths failed");
```

Comment:
That private looks funny, ask the stream what is the root in order to create the path.